File: adversarial-harness/python/policy_engine/policy_timing_envelope.py
"""Policy evaluation timing envelope — mirrors policy-timing-envelope.ts."""

from __future__ import annotations

import os
import time


def _env_int(name: str, default: int) -> int:
    raw = os.environ.get(name)
    if raw is None:
        return default
    try:
        return max(0, int(raw))
    except ValueError:
        return default


def is_policy_timing_envelope_enabled() -> bool:
    return os.environ.get("GUARDIAN_POLICY_TIMING_ENVELOPE") != "false"


def policy_min_eval_ms() -> int:
    return _env_int("MCP_GUARDIAN_POLICY_MIN_EVAL_MS", 25)


def wait_policy_timing_envelope_sync(started_at: float) -> None:
    if not is_policy_timing_envelope_enabled():
        return
    min_ms = policy_min_eval_ms()
    deadline = started_at + min_ms / 1000.0
    while time.time() < deadline:
        pass


def wait_policy_timing_envelope_async(started_at: float) -> None:
    if not is_policy_timing_envelope_enabled():
        return
    min_ms = policy_min_eval_ms()
    elapsed_ms = (time.time() - started_at) * 1000.0
    if elapsed_ms < min_ms:
        time.sleep((min_ms - elapsed_ms) / 1000.0)

File: src/policy/policy-timing-envelope.ts
/**
 * Policy evaluation timing envelope — normalizes wall-clock latency so pass vs block
 * decisions do not leak via measurable timing differences (within Guardian's boundary).
 */
import { envInt } from '../utils/eval-bounds.js';

export function isPolicyTimingEnvelopeEnabled(): boolean {
  return process.env['GUARDIAN_POLICY_TIMING_ENVELOPE'] !== 'false';
}

export function policyMinEvalMs(): number {
  return envInt('MCP_GUARDIAN_POLICY_MIN_EVAL_MS', 25);
}

export function proxyTimingNormalizeMs(): number {
  return envInt('MCP_GUARDIAN_PROXY_TIMING_NORMALIZE_MS', 15);
}

/** Synchronous minimum duration (used by harness sync evaluate). */
export function waitPolicyTimingEnvelopeSync(startedAt: number): void {
  if (!isPolicyTimingEnvelopeEnabled()) return;
  const minMs = policyMinEvalMs();
  const deadline = startedAt + minMs;
  while (Date.now() < deadline) {
    /* intentional spin — bounds policy path timing leakage */
  }
}

/** Async minimum duration (production proxy evaluateAsync). */
export async function waitPolicyTimingEnvelopeAsync(startedAt: number): Promise<void> {
  if (!isPolicyTimingEnvelopeEnabled()) return;
  const minMs = policyMinEvalMs();
  const elapsed = Date.now() - startedAt;
  if (elapsed < minMs) {
    await new Promise((resolve) => setTimeout(resolve, minMs - elapsed));
  }
}

/** Optional extra delay on tools/call to reduce request-path timing oracle. */
export async function waitProxyTimingNormalize(startedAt: number): Promise<void> {
  if (process.env['GUARDIAN_PROXY_TIMING_NORMALIZE'] === 'false') return;
  const minMs = proxyTimingNormalizeMs();
  const elapsed = Date.now() - startedAt;
  if (elapsed < minMs) {
    await new Promise((resolve) => setTimeout(resolve, minMs - elapsed));
  }
}
